### ISSUE TITLE
[Security] Kong still connected to original docker container ip after restarting scheduler

### ISSUE THREAD

**Issue by cherrycl** [edgexfoundry/edgex-taf#403]
On TAF test case [IntervalactionGET006](https://github.com/edgexfoundry/edgex-taf/blob/33797ce8e61e0a93c7ee0819ab9a239f5ed4ef62/TAF/testScenarios/functionalTest/V2-API/support-scheduler/intervalaction/GET.robot#L70), the `support-scheduler`is restarted after updating `Consul` configuration.
After restarting the scheduler, the test always got failure with error `"message":"An invalid response was received from the upstream server"`.

On kong log, found the following error.
`2021/05/06 08:28:38 [error] 113#0: *536 upstream timed out (110: Operation timed out) while connecting to upstream, client: 192.168.240.1, server: kong, request: "POST /scheduler/api/v2/interval HTTP/1.1", upstream: "http://192.168.240.11:48085/api/v2/interval", host: "localhost:8443"`

But After restarting the scheduler, the docker container IP is changed to `192.168.240.9`
```
cherry@cherry-snapshot:~/edgexfoundry$ docker inspect edgex-support-scheduler | grep IPAddress
            "SecondaryIPAddresses": null,
            "IPAddress": "",
                    "IPAddress": "192.168.240.9",
```
After 10 minutes, kong connected to the correct `scheduler` container IP. 

Running full TAF test always got this error. If only run the case `IntervalactionGET006`, this error might not happen.
This issue only occurred on `x86_64` platform.

**Comment by cherrycl** [comment]
@beaufrusetta @lenny-intel Do you have any idea for this error?

**Comment by lenny-intel** [comment]
@hutchic , here is the Kong issue we discussed yesterday that you were wanting to recreate. 
@cherrycl , can this test need to be add to the SmokeTests for edgex-compose so we can play with version of Kong to see if it has been resolved in newer version? THX!

**Comment by lenny-intel** [comment]
@cherrycl , FYI we have updated the 3rd party dependencies to their latest versions, which includes Kong to 2.4. 
https://github.com/edgexfoundry/edgex-compose/pull/78
Please let us know if this has any impact on this issue. THX!

**Comment by cherrycl** [comment]
> @hutchic , here is the Kong issue we discussed yesterday that you were wanting to recreate.
> @cherrycl , can this test need to be add to the SmokeTests for edgex-compose so we can play with version of Kong to see if it has been resolved in newer version? THX!

@lenny-intel Only add the IntervalactionGET006 can't recreate the issue, so I added the whole suite to the SmokeTest in PR #404.

**Comment by cherrycl** [comment]
> @cherrycl , FYI we have updated the 3rd party dependencies to their latest versions, which includes Kong to 2.4.
> [edgexfoundry/edgex-compose#78](https://github.com/edgexfoundry/edgex-compose/pull/78)
> Please let us know if this has any impact on this issue. THX!

@lenny-intel I raised an issue [edgexfoundry/edgex-compose #81](https://github.com/edgexfoundry/edgex-compose/issues/81) for the `mosquitto v2` issue. Would you please help to take care of it?

**Comment by lenny-intel** [comment]
@cherrycl , can these scheduler tests be removed for the Secure mode tests until we resolve the issue with Kong?

**Comment by cherrycl** [comment]
The related tests are removed from the smoke test.

**Comment by lenny-intel** [comment]
> The related tests are removed from the smoke test.

@cherrycl , thanks! This still causing the full TAF suite to fail, correct?

**Comment by cherrycl** [comment]
> > The related tests are removed from the smoke test.
> 
> @cherrycl , thanks! This still causing the full TAF suite to fail, correct?

@lenny-intel No, it only causes a few tests to fail. For recreating the issue, it needs to run [the full suite](https://github.com/edgexfoundry/edgex-taf/blob/33797ce8e61e0a93c7ee0819ab9a239f5ed4ef62/TAF/testScenarios/functionalTest/V2-API/support-scheduler/intervalaction/GET.robot), but it only causes the tests to fail after the test `IntervalactionGET006`.

**Comment by hutchic** [comment]
```
git clone git@github.com:edgexfoundry/edgex-compose.git
cd edgex-compose
make run
docker-compose -p edgex ps #wait for healthy
make get-token
export token=TOKENFROMABOVE
curl -H "Authorization: Bearer $token" --insecure https://localhost:8443/scheduler/api/v1/ping #works
docker-compose -p edgex restart scheduler
curl -H "Authorization: Bearer $token" --insecure https://localhost:8443/scheduler/api/v1/ping #if scheduler changed IP's this fails
```

Kong logs
```
kong                     | 2021/05/12 17:26:48 [error] 106#0: *550 connect() failed (113: Host is unreachable) while connecting to upstream, client: 172.23.0.1, server: kong, request: "GET /scheduler/api/v1/ping HTTP/2.0", upstream: "http://172.23.0.11:48085/api/v1/ping", host: "localhost:8443"
kong                     | 2021/05/12 17:26:51 [error] 106#0: *550 connect() failed (113: Host is unreachable) while connecting to upstream, client: 172.23.0.1, server: kong, request: "GET /scheduler/api/v1/ping HTTP/2.0", upstream: "http://172.23.0.11:48085/api/v1/ping", host: "localhost:8443"
```

Expected result: it proxies to the new IP
Actual result: some of the nginx workers seem to continue to try to proxy to the old ip

**Comment by cherrycl** [comment]
Was this issue fixed? The related tests passed on [Jenkins functional-test #451](https://jenkins.edgexfoundry.org/view/EdgeX%20Foundry%20Project/job/edgexfoundry/job/edgex-taf-pipelines/job/functional-test/451/console).

**Comment by cherrycl** [comment]
> Was this issue fixed? The related tests passed on [Jenkins functional-test #451](https://jenkins.edgexfoundry.org/view/EdgeX%20Foundry%20Project/job/edgexfoundry/job/edgex-taf-pipelines/job/functional-test/451/console).

I also verified on my local env. All tests are passed.

**Comment by cherrycl** [comment]
The issue happened again when using the latest edgex image.

**Comment by lenny-intel** [comment]
@cherrycl , nothing has been changed to fix this issue. I suspect that it will randomly pass due to the restart not causing a new IP.

**Comment by cloudxxx8** [comment]
@cherrycl does this issue only happen in x86?

**Comment by cherrycl** [comment]
> @cherrycl does this issue only happen in x86?

@cloudxxx8 It seems also happened on arm64, but arm64 only failed once so far.

**Comment by lenny-intel** [comment]
@cherrycl , @cloudxxx8 , @JamesKButcher , @jpwhitemn , One possible work around for TAF testing is to immediately restart the scheduler container once all service have started. Then have the Scheduler test toward the end of testing so there is time for Kong cache of original IP to timeout and use the new IP. This assumes no other containers are restarted in between so that the scheduler container will get same IP when it is started as part of the existing test.

This is based on the the fact that once the container is restarted and receives new IP, it always gets the same IP on subsequent restarts as long as no other containers are restarted.

**Comment by cloudxxx8** [comment]
@lenny-intel I just discussed with Cherry.  The purpose of this particular test case is to verify the behavior after modifying a non-writable configuration.  It requires the service restarting, and the test should be atomic.  It's hard to diagnose the root cause in the future we workaround by the sequence change.  The simple way to avoid this failure is to disable this test case in the security enablement.

**Comment by lenny-intel** [comment]
@cloudxxx8 , @jpwhitemn , this makes sense for this test since it isn't about testing access via the API Gateway, so disabling it for security enablement makes sense.

**Comment by hutchic** [comment]
For the purposes of testing we can set [dns_valid_ttl](https://docs.konghq.com/gateway-oss/2.4.x/configuration/#dns_valid_ttl) to a sufficiently low value

**Comment by cloudxxx8** [comment]
@hutchic sounds like a good idea.  we can set it to 1 second, and sleep 1 second after the service restarting.

**Comment by cloudxxx8** [comment]
@lenny-intel we verified this change works in our local environment
should we add an environment variable `KONG_DNS_VALID_TTL=1` to `kong` in add-security.yml or in the taf folder only?

**Comment by lenny-intel** [comment]
@cloudxxx8 , yes, I have that WIP. But still see issues on my system that I need to verify.

**Comment by cherrycl** [comment]
Issue fixed by [edgexfoundry/edgex-compose PR #116](https://github.com/edgexfoundry/edgex-compose/issues/116).
